### PR TITLE
[MkFit] Tracker geometry extraction from Reco geom modules

### PR DESCRIPTION
**Pull request by osschar**
This PR addresses the first long-term item in cms-sw#36966 - removal of hardcoded phasI geometry from mkFit and implementation of proper geometry extraction from RECO geometry loaded in current cmsRun process.

* Extract mkFit layer r-z bounding regions and TEC r-coverage gaps from individual modules.
* Introduce ModuleInfo with module position, normal and principal direction. Each layer holds a std::vector and a mapping from detId to short in-layer module index that was constructed on the fly beforehand.
* LayerInfo: remove/cleanup is_pixb/pixe/tib/tob/tid/tec data members and functions. They were not really used. Put in 'int subdet' that can be used to the same effect in scoring / cleaning functions.
* Remove geometry creation files with hardcoded geometry information: MkFit/plugins/createPhase1TrackerGeometry.h/cc/acc

Corresponding changes for standalone extras are on https://github.com/trackreco/mkFit-external/tree/geom-extract

### PR validation:
* TTbar + < PU >=50: http://uaf-10.t2.ucsd.edu/~mmasciov/MIC/rescaleError100/MTV_TTbarPU50_extractedGeometry_mkFitOnly/
* 10μ, pT=0.6 GeV, 1.5<|η|<2.5: http://uaf-10.t2.ucsd.edu/~mmasciov/MIC/rescaleError100/MTV_10muPt0p6Eta1p5to2p5_extractedGeometry_mkFitOnly/

There are minimal changes in efficiency and fake rates, below 1% level. These are expected as the hardcoded mkFit geometry was not identical to one obtained by proper extraction. There are differences in extent of endcap r-gaps (O(1cm)) and z-extent of endcap mkFit layers (O(5cm)).

Timing is not expected to change significantly (if at all) as this PR only affects extents of mkFit layers.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc [hunk 1]
```diff
@@ -10,44 +11,430 @@
...
 
+//------------------------------------------------------------------------------
+
+#include "DataFormats/GeometrySurface/interface/RectangularPlaneBounds.h"
+#include "DataFormats/GeometrySurface/interface/TrapezoidalPlaneBounds.h"
+//#include "Geometry/CommonDetUnit/interface/GlobalTrackingGeometry.h"
+//#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+//#include "Geometry/TrackerGeometryBuilder/interface/StripGeomDetUnit.h"
+//#include "Geometry/TrackerGeometryBuilder/interface/StripGeomDetType.h"
+//#include "Geometry/TrackerGeometryBuilder/interface/RectangularPixelTopology.h"
```
Comment (jpata) [inline RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc:30]: any reason to keep these around?

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc [hunk 2]
```diff
@@ -10,44 +11,430 @@
...
+  return false;
+}
+
+void MkFitGeometryESProducer::GapCollector::print_gaps() {
+  auto i = m_coverage.begin();
+  while (i != m_coverage.end()) {
+    auto j = i;
+    ++j;
+    if (j != m_coverage.end()) {
+      printf("(%f, %f)->%f ", i->y, j->x, j->x - i->y);
```
Comment (jpata) [inline RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc:166]: LogVerbatim perhaps?

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc [hunk 3]
```diff
@@ -10,44 +11,430 @@
...
+                            info.shape[4],  //dy2
+                            info.shape[1],  //dx3
+                            info.shape[2],  //dx4
+                            0);             //alpha2
+
+      TGeoTrap::TGeoTrap(Double_t dz, Double_t theta, Double_t phi, Double_t h1,
+               Double_t bl1, Double_t tl1, Double_t alpha1, Double_t h2, Double_t bl2,
+               Double_t tl2, Double_t alpha2)
+          :TGeoArb8("", 0, 0)
+    {
```
Comment (jpata) [inline RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc:222]: is this intended as documentation, or is it leftover?

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc [hunk 4]
```diff
@@ -10,44 +11,430 @@
...
+                                         << " width = " << itvl.y - itvl.x;
+      trk_info.layer_nc(layer).set_r_hole_range(itvl.x, itvl.y);
+    }
+  }
+}
+
+//------------------------------------------------------------------------------
+// clang-format off
+namespace {
+  const float phase1QBins[] = {
```
Comment (jpata) [inline RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc:383]: are these copied from somewhere else? perhaps I miss the point, but isn't this still hardcoding?

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc [hunk 5]
```diff
@@ -10,44 +11,430 @@
...
+//------------------------------------------------------------------------------
+
+#include "DataFormats/GeometrySurface/interface/RectangularPlaneBounds.h"
+#include "DataFormats/GeometrySurface/interface/TrapezoidalPlaneBounds.h"
+//#include "Geometry/CommonDetUnit/interface/GlobalTrackingGeometry.h"
+//#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+//#include "Geometry/TrackerGeometryBuilder/interface/StripGeomDetUnit.h"
+//#include "Geometry/TrackerGeometryBuilder/interface/StripGeomDetType.h"
+//#include "Geometry/TrackerGeometryBuilder/interface/RectangularPixelTopology.h"
+#include <fstream>
```
Comment (jpata) [inline RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc:31]: what are you using from this include?

File: RecoTracker/MkFitCore/src/TrackerInfo.cc [hunk 6]
```diff
@@ -11,12 +12,39 @@ namespace mkfit {
...
+
   void LayerInfo::set_r_hole_range(float rh1, float rh2) {
     m_has_r_range_hole = true;
     m_hole_r_min = rh1;
     m_hole_r_max = rh2;
   }
 
+  void LayerInfo::print_layer() const {
+    // clang-format off
+    printf("Layer %2d  r(%7.4f, %7.4f) z(% 9.4f, % 9.4f) is_brl=%d, is_pix=%d, is_stereo=%d, q_bin=%.2f\n",
```
Comment (jpata) [inline RecoTracker/MkFitCore/src/TrackerInfo.cc:39]: should these also be LogVerbatim?

File: RecoTracker/MkFitCore/src/TrackerInfo.cc [hunk 7]
```diff
@@ -44,19 +72,192 @@ namespace mkfit {
...
+      int f_n_layers = -1;
+
+      GeomFileHeader() = default;
+
+      constexpr static int s_magic = 0xB00F;
+      constexpr static int s_version = 1;
+    };
+
+    template <typename T>
+    int write_std_vec(FILE* fp, const std::vector<T>& vec, int el_size = 0) {
```
Comment (jpata) [inline RecoTracker/MkFitCore/src/TrackerInfo.cc:115]: can you actually unpack a bit the purpose of this binary file? Is it for private/standalone debugging, or something intended to be used in the official development process?
I could be a little concerned about the long-term maintainability in the latter case.

**Review by jpata — COMMENTED** [review]
Thanks for the follow-up. Some minor code comments inline.

I didn't notice anything significant in the tracking validation plots you linked, but perhaps experts (@cms-sw/tracking-pog-l2) could take a look too.

Since mkFit is now running in production, would it make sense to do a throughput check here (or perhaps starting from just igprof), just to make sure we don't have any unexpected surprises?

File: RecoTracker/MkFit/test/DumpMkFitGeometry.cc [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+#include "RecoTracker/Record/interface/TrackerRecoGeometryRecord.h"
+
+#include "RecoTracker/MkFit/interface/MkFitGeometry.h"
+
+#include "RecoTracker/MkFitCore/interface/TrackerInfo.h"
+
+class DumpMkFitGeometry : public edm::one::EDAnalyzer<> {
+public:
+  explicit DumpMkFitGeometry(const edm::ParameterSet& config);
+  ~DumpMkFitGeometry(void) override {}
```
Comment (mmusich) [inline RecoTracker/MkFit/test/DumpMkFitGeometry.cc:16]: could be `default`.

File: RecoTracker/MkFit/test/dumpMkFitGeometry.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+process.add_(cms.ESProducer("MkFitGeometryESProducer"))
+
+defaultOutputFileName="phase1-trackerinfo.bin"
+
+# level: 0 - no printout; 1 - print layers, 2 - print modules
+# outputFileName: binary dump file; no dump if empty string
+process.dump = cms.EDAnalyzer("DumpMkFitGeometry",
+                              level   = cms.untracked.int32(1),
+                       outputFileName = cms.untracked.string(defaultOutputFileName)
```
Comment (mmusich) [inline RecoTracker/MkFit/test/dumpMkFitGeometry.py:31]: indentation looks funny.

**Review by mmusich — COMMENTED** [review]
cosmetic comments, feel free to ignore, if the files won't be updated further.

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc (hunk 1 above)
Comment (osschar) [reply RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc:30]: I'll remove them, I was basing this on Fireworks Reco geometry extractor and there we also extract "topology".

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc (hunk 2 above)
Comment (osschar) [reply RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc:166]: I can look it up and replace ... but this only called for crosscheck when tuning the ignored gap-size when r-coverage is spotty. I suspect I'll only need it when working on the phaseII geometry, if it also has a coverage gap :)

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc (hunk 3 above)
Comment (osschar) [reply RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc:222]: It's documentation ... it was quite fun to figure out how to interpret the Bounds objkect parameter array. I can move it to a file in doc/ and reference it.

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc (hunk 4 above)
Comment (osschar) [reply RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc:383]: Well, that's all that's left, there were many many more before, you know :)

These determine how hits are binned in z (barrel) and r (endcap) for fast lookup. In principle it's determined from module extent in that direction so we should be able to extract this during the loop over the modules as well. But this is potentially a larger physics change and might require also retuning of hit selection window parameters.

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc (hunk 5 above)
Comment (osschar) [reply RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc:31]: will remove if not used.

File: RecoTracker/MkFitCore/src/TrackerInfo.cc (hunk 6 above)
Comment (osschar) [reply RecoTracker/MkFitCore/src/TrackerInfo.cc:39]: These are also only used when making a new geometry / debugging ... and those are, however, also used in standalone.

File: RecoTracker/MkFitCore/src/TrackerInfo.cc (hunk 7 above)
Comment (osschar) [reply RecoTracker/MkFitCore/src/TrackerInfo.cc:115]: This is for the geometry to be picked up in standalone.

File: RecoTracker/MkFit/test/DumpMkFitGeometry.cc (hunk 8 above)
Comment (osschar) [reply RecoTracker/MkFit/test/DumpMkFitGeometry.cc:16]: ok, will replace.

File: RecoTracker/MkFit/test/DumpMkFitGeometry.cc [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+private:
+  void analyze(const edm::Event& event, const edm::EventSetup& eventSetup) override;
+
+  edm::ESGetToken<MkFitGeometry, TrackerRecoGeometryRecord> m_mkfGeoToken;
+
+  int m_level;
+  std::string m_outputFileName;
```
[the remark this answers is not in the document]
Comment (osschar) [reply RecoTracker/MkFit/test/DumpMkFitGeometry.cc:24]: Hmmh, Slava told me m_ is also ok -- but I can replace it if you prefer. This is in test/ dir, just for dumping the geometry into binary file for standalone use.

File: RecoTracker/MkFit/test/dumpMkFitGeometry.py (hunk 9 above)
Comment (osschar) [reply RecoTracker/MkFit/test/dumpMkFitGeometry.py:31]: Sorry, how is it supposed to look? :)

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc [hunk 11]
```diff
@@ -10,44 +11,430 @@
...
+  void addTOBGeometry(mkfit::TrackerInfo &trk_info);
+  void addTIDGeometry(mkfit::TrackerInfo &trk_info);
+  void addTECGeometry(mkfit::TrackerInfo &trk_info);
+
   edm::ESGetToken<TrackerGeometry, TrackerDigiGeometryRecord> geomToken_;
-  edm::ESGetToken<GeometricSearchTracker, TrackerRecoGeometryRecord> trackerToken_;
   edm::ESGetToken<TrackerTopology, TrackerTopologyRcd> ttopoToken_;
+  edm::ESGetToken<GeometricSearchTracker, TrackerRecoGeometryRecord> trackerToken_;
+
+  const TrackerTopology *m_trackerTopo = nullptr;
```
[the remark this answers is not in the document]
Comment (osschar) [reply RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc:80]: ok, will fix this one and the one in test/ :)

**Comment by osschar** [comment]
@slava77, @jpata asked for basic igProf measurement / comparison ... can this be done through github ifc?

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc (hunk 1 above)
Comment (osschar) [reply RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc]: removed

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc (hunk 3 above)
Comment (osschar) [reply RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc]: moved to doc/reco-geom-notes.txt

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc (hunk 5 above)
Comment (osschar) [reply RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc]: removed

File: RecoTracker/MkFit/test/DumpMkFitGeometry.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (osschar) [reply RecoTracker/MkFit/test/DumpMkFitGeometry.cc]: renamed

File: RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (osschar) [reply RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc]: renamed

File: RecoTracker/MkFit/test/DumpMkFitGeometry.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (slava77) [reply RecoTracker/MkFit/test/DumpMkFitGeometry.cc]: that was my practice for 8-9 years of direct reviews in the github; just a clear pattern different from local variable name was required.
Note that we have O(10K) lines of already accepted code in mkFit where `m_` was accepted as OK just recently.

**Comment by slava77** [comment]
> @slava77, @jpata asked for basic igProf measurement / comparison ... can this be done through github ifc?

I will add `enable profiling`.

I'm just not sure what to look for in this context: small variations in tracks are hard to interpret; so I haven't added this part in the initial `please test`

**Comment by slava77** [comment]
enable profiling

File: RecoTracker/MkFit/test/DumpMkFitGeometry.cc (hunk 8 above)
Comment (osschar) [reply RecoTracker/MkFit/test/DumpMkFitGeometry.cc]: removed

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
> **Additional Tests**: PROFILING

CPU
- `MkFitProducer` on 20 events [7.02%/24.79](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_4_X_2022-04-01-2300/slc7_amd64_gcc10/profiling/11834.21/igprofCPU_step3/31) -> [6.94%/24.77](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_4_X_2022-04-01-2300/slc7_amd64_gcc10/profiling/11834.21/PR-520b22/23613/igprofCPU_step3/31) seems like a no-change
-  ES producer calls are registered, but are near 0 both [before](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_4_X_2022-04-01-2300/slc7_amd64_gcc10/profiling/11834.21/igprofCPU_step3.1/10) and [after](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_4_X_2022-04-01-2300/slc7_amd64_gcc10/profiling/11834.21/PR-520b22/23613/igprofCPU_step3.1/10)

Memory
- mkFit geometry [575,420](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_4_X_2022-04-01-2300/slc7_amd64_gcc10/profiling/11834.21/igprofMEM_step3.1/2978) -> [1,389,137](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_4_X_2022-04-01-2300/slc7_amd64_gcc10/profiling/11834.21/PR-520b22/23613/igprofMEM_step3.1/2065) should correspond to more per-module data; although I didn't try to do the bean counting.

So, this looks somewhat as expected.

**Comment by osschar** [comment]
Yes, we have O(22k) modules with 40 bytes per module -> 22000*(3*3*4 + 4) = 880000

**Comment by jpata** [comment]
+reconstruction
- improves geometry handling in mkFit
- many small changes to tracking and everything downstream, but according to the tracking validation in the PR description, no physics changes are expected
- CPU and memory do not look to be significantly affected in the igprof test

**Comment by osschar** [comment]
Rats, I forgot include MkFit/test/BuildFile.xml for building of the standalone geometry dumper, MkFit/test/DumpMkFitGeometry.cc.

Shall I put it in here or I wait until the next PR so we don't need to go through test etc. hooplas again? The file is rather trivial:
```
matevz@phi3 RecoTracker> cat MkFit/test/BuildFile.xml
<library file="DumpMkFitGeometry.cc" name="RecoTrackerMkFitTest">
  <flags EDM_PLUGIN="1"/>
  <use name="FWCore/Framework"/>
  <use name="RecoTracker/MkFit"/>
  <use name="RecoTracker/MkFitCore"/>
  <use name="RecoTracker/Record"/>
</library>
```

**Comment by qliphy** [comment]
@osschar You may add it within this PR, and can also take the chance to implement 
https://github.com/cms-sw/cmssw/pull/37418/files#r839654493

**Comment by osschar** [comment]
Naaah, just go ahead and merge, I'll fix that print in another PR as well, it's almost ready -- the second long term item from our todo list, removal of Ice/ and consolidation of radix sort usage.

**Comment by qliphy** [comment]
+1
to follow up on https://github.com/cms-sw/cmssw/pull/37418#issuecomment-1089602049

**Comment by osschar** [comment]
Thank you! :)

**Comment by mmusich** [comment]
For the record https://github.com/cms-sw/cmssw/pull/37418#issuecomment-1089262916 has been fixed at https://github.com/cms-sw/cmssw/pull/37586

**Pull request merged**